File: plugins/plugin-evm/python/elizaos_plugin_evm/actions/gov_vote.py
import logging

from elizaos_plugin_evm.error import EVMError
from elizaos_plugin_evm.providers.wallet import EVMWalletProvider
from elizaos_plugin_evm.types import VoteParams

logger = logging.getLogger(__name__)


async def execute_vote(
    provider: EVMWalletProvider,
    params: VoteParams,
) -> str:
    logger.info(
        "Executing governance vote on proposal %s on governor %s on %s",
        params.proposal_id,
        params.governor,
        params.chain.value,
    )

    if params.support not in [0, 1, 2]:
        raise EVMError.invalid_params("Support must be 0 (Against), 1 (For), or 2 (Abstain)")

    from web3 import Web3

    w3 = Web3()
    vote_selector = w3.keccak(text="castVote(uint256,uint8)")[:4]
    encoded_params = w3.codec.encode(
        ["uint256", "uint8"],
        [int(params.proposal_id), params.support],
    )

    data = vote_selector.hex() + encoded_params.hex()

    tx_hash = await provider.send_transaction(
        chain=params.chain,
        to=params.governor,
        value=0,
        data=data,
    )

    await provider.wait_for_transaction(params.chain, tx_hash)

    vote_type = {0: "Against", 1: "For", 2: "Abstain"}[params.support]
    logger.info("Vote cast (%s): %s", vote_type, tx_hash)
    return tx_hash


vote_action = {
    "name": "VOTE",
    "description": "Cast a vote on a DAO governance proposal",
    "similes": [
        "GOVERNANCE_VOTE",
        "cast_vote",
        "vote_proposal",
    ],
    "examples": [
        "Vote for proposal 1 on governor 0x1234 on Ethereum",
        "Vote against proposal 42 on the DAO at 0xabcd on Base",
    ],
    "handler": execute_vote,
}

File: plugins/plugin-evm/typescript/actions/gov-vote.ts
import type { ActionResult, HandlerCallback, IAgentRuntime, Memory, State } from "@elizaos/core";
import { type Address, encodeFunctionData, type Hex } from "viem";
import governorArtifacts from "../contracts/artifacts/OZGovernor.json";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { WalletProvider } from "../providers/wallet";
import { voteTemplate } from "../templates";
import type { SupportedChain, Transaction, VoteParams } from "../types";

export { voteTemplate };

const spec = requireActionSpec("VOTE_PROPOSAL");

export class VoteAction {
  constructor(private walletProvider: WalletProvider) {}

  async vote(params: VoteParams): Promise<Transaction> {
    const walletClient = this.walletProvider.getWalletClient(params.chain);

    if (!walletClient.account) {
      throw new Error("Wallet account is not available");
    }

    const proposalId = BigInt(params.proposalId.toString());
    const support = BigInt(params.support.toString());

    const txData = encodeFunctionData({
      abi: governorArtifacts.abi,
      functionName: "castVote",
      args: [proposalId, support],
    });

    try {
      const chainConfig = this.walletProvider.getChainConfigs(params.chain);
      const publicClient = this.walletProvider.getPublicClient(params.chain);

      // @ts-expect-error - viem type narrowing issue with sendTransaction parameters
      const hash = await walletClient.sendTransaction({
        account: walletClient.account,
        to: params.governor,
        value: BigInt(0),
        data: txData as Hex,
        chain: chainConfig,
      });

      const receipt = await publicClient.waitForTransactionReceipt({
        hash,
      });

      return {
        hash,
        from: walletClient.account?.address as `0x${string}`,
        to: params.governor,
        value: BigInt(0),
        data: txData as Hex,
        chainId: this.walletProvider.getChainConfigs(params.chain).id,
        logs: receipt.logs,
      };
    } catch (error: unknown) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      throw new Error(`Vote failed: ${errorMessage}`);
    }
  }
}

export const voteAction = {
  name: spec.name,
  description: spec.description,
  handler: async (
    runtime: IAgentRuntime,
    _message: Memory,
    _state: State,
    options: Record<string, unknown>,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    try {
      if (!options.chain || !options.governor || !options.proposalId || !options.support) {
        throw new Error("Missing required parameters for vote");
      }

      const voteParams: VoteParams = {
        chain: options.chain as SupportedChain,
        governor: options.governor as Address,
        proposalId: String(options.proposalId),
        support: Number(options.support),
      };

      const privateKey = runtime.getSetting("EVM_PRIVATE_KEY") as `0x${string}`;
      const walletProvider = new WalletProvider(privateKey, runtime);
      const action = new VoteAction(walletProvider);
      const result = await action.vote(voteParams);
      return {
        success: true,
        text: `Vote submitted successfully. Transaction hash: ${result.hash}`,
        data: {
          transactionHash: result.hash,
          from: result.from,
          to: result.to,
          chain: voteParams.chain,
        },
      };
    } catch (error: unknown) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      if (callback) {
        callback({ text: `Error: ${errorMessage}` });
      }
      return {
        success: false,
        text: `Error: ${errorMessage}`,
      };
    }
  },
  template: voteTemplate,
  validate: async (runtime: IAgentRuntime) => {
    const privateKey = runtime.getSetting("EVM_PRIVATE_KEY");
    return typeof privateKey === "string" && privateKey.startsWith("0x");
  },
  examples: [
    [
      {
        name: "user",
        content: {
          text: "Vote yes on proposal 123 on the governor at 0x1234567890123456789012345678901234567890 on Ethereum",
          action: "GOVERNANCE_VOTE",
        },
      },
    ],
    [
      {
        name: "user",
        content: {
          text: "Vote no on proposal 456 on the governor at 0xabcdef1111111111111111111111111111111111 on Ethereum",
          action: "GOVERNANCE_VOTE",
        },
      },
    ],
    [
      {
        name: "user",
        content: {
          text: "Abstain from voting on proposal 789 on the governor at 0x0000111122223333444455556666777788889999 on Ethereum",
          action: "GOVERNANCE_VOTE",
        },
      },
    ],
  ],
  similes: spec.similes ? [...spec.similes] : [],
};
